lib/irb.rb

```diff
@@ -1523,6 +1528,8 @@ def irb(show_code: true)
     debugger_irb = IRB.instance_variable_get(:@debugger_irb)
 
     irb_path = File.expand_path(source_location[0])
+    # We need to change the irb_path to distinguish source_location of method defined in the actual file and method defined in irb session.
+    irb_path = "#{irb_path}(#{IRB.conf[:IRB_NAME]})" if File.exist?(irb_path)
```

Comment:
If we don't append `"(#{irb_name})"`, show_source won't work correctly.

Save this to `a.rb`
```ruby
def bar; end
binding.irb
```
And run `ruby -I path/to/irb/lib a.rb`
```ruby
irb(main):001> def foo; end
irb(main):001> $ foo
```
Without appending, `method(:foo).source_location` and `method(:bar).source_location` will both return the same value `["a.rb", 1]`

Reply:
Because the `edit` command uses `context.irb_path` to open the breakpoint's file, this will break it.


```
irb(main):001> edit
command: 'code'
   path: /Users/hung-wulo/src/github.com/ruby/irb/test.rb
=> true
```

With the PR:

```
From: test.rb @ line 7 :

    2:   def bar
    3:     puts '123'
    4:   end
    5: end
    6: 
 => 7: binding.irb

irb(main):001> edit
/Users/hung-wulo/src/github.com/ruby/irb/lib/irb/source_finder.rb:35:in `eval': (eval):1: unknown regexp options - hg (SyntaxError)
/Users/hung-wulo/src/github.com/ruby/irb...
```

So instead of changing the value of `context#irb_path`, let's modify the one we pass to `@workspace.evaluate`? I feel it's a better solution even without the issue as it makes the in-session definitions have clearer source location too.

Reply:
Nice catch! changed to:
```ruby
eval_path = File.exist?(irb_path) ? "#{irb_path}(#{IRB.conf[:IRB_NAME]})" : irb_path
...
result = @workspace.evaluate(line, eval_path, line_no)
```

I also changed (https://github.com/ruby/irb/pull/862/commits/1ec0b06d13f9ac7411d90249f32f953a7785f4c5) edit and add a test for non-existing irb_path for:
```
% ruby -I path/to/irb/lib a.rb
From: a.rb @ line 1 :

 => 1: binding.irb

irb(main):001> IRB.CurrentContext.irb_path
=> "/Users/tomoya.ishida/github/ruby/irb/a.rb"
irb(main):002> edit
command: 'emacs'
   path: /Users/tomoya.ishida/github/ruby/irb/a.rb
=> true
irb(main):003> binding.irb # this part is evaluated with eval_path="a.rb(irb)"
irb(main):001> IRB.CurrentContext.irb_path
=> "/Users/tomoya.ishida/github/ruby/irb/a.rb(irb)"
irb(main):002> edit
Can not find file: /Users/tomoya.ishida/github/ruby/irb/a.rb(irb)
=> nil
```
To avoid path passed to SourceFinder (SourceFinder will evaluate path as ruby code)
```
irb(main):002> edit
`eval': unknown regexp options - tya (SyntaxError)
/Users/tomoya.ishida/github/ruby/irb/a
      ^~~~~~~
```